### PR TITLE
FO: override a method uses private variable when you inheritance CustomerAddressFormatterCore

### PR DESCRIPTION
**Pull request by azisyus**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | this PR  solves same thing in different class, check #9417  
| Type?         | improvement
| Category?     | FO 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | these access modifiers cant break anything

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9420)

### CODE REVIEW

**Review by Quetzacoalt91 — APPROVED** [review]

File: classes/form/CustomerAddressFormatter.php [hunk 1]
```diff
@@ -29,10 +29,10 @@
 
 class CustomerAddressFormatterCore implements FormFormatterInterface
 {
-    private $country;
-    private $translator;
-    private $availableCountries;
-    private $definition;
+    protected $country;
```
Comment (mickaelandrieu) [inline classes/form/CustomerAddressFormatter.php:32]: You don't need this as you have getter and setter for `$country` and others come from the constructor.

If you need to alter or retrieve `$availableCountries`, do it before the instantiation of this object, same for the translator.

@Quetzacoalt91 making these properties protected force us to keep them for backward compatibility with all classes that may extend it. As there is an interface, the right way for developers to alter the behavior is to create their own implementation, not to extends the current one.

I'm more in favor of making this class `final` in 1.8 than allow people to alter it and add it in yet another `overrides` folder: wdyt?

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]
For me it's wrong to do this, and this will encourage class overriding, what we don't want (plus, this class have an interface so people are already able to create their own implementation the right way).

File: classes/form/CustomerAddressFormatter.php (hunk 1 above)
Comment (mickaelandrieu) [reply classes/form/CustomerAddressFormatter.php:32]: The only boring thing is this line:

```php
         $this->definition = Address::$definition['fields'];
```

We may inject definition instead ...

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]
I understand the need, but not the way to do it.

**Comment by mickaelandrieu** [comment]
Hello @azisyus,

can we discuss a Hook that can help you with what you want to do?

**Comment by azisyus** [comment]
@mickaelandrieu actually and basically i need 2 different address form, i mean in original definition of address form 

`Address::$definition['fields'];`

 has company and vat_number fields and in my country while buying new things from ecom sites we dont need these fields for **delivery address** form but same time i need these fields for **invoice address** form because of laws then i though, there is only one table i can store addresses so i can show different parts of address form in different places and it will look like a 2 different form :D it was tricky and disgusting way and end of the week i just rolled back many commit and put the original form to everywhere and i need a hook way to do it or i need to override all classes got relation with addresses

**Comment by mickaelandrieu** [comment]
maybe this demonstration module can help? https://github.com/mickaelandrieu/products/blob/master/products.php#L110

Basically, you can [alter](https://github.com/mickaelandrieu/products/blob/master/src/AlternativeDescription.php#L24) an Object Model definition without an override or update of its file, then you could use a hook and decide in which cases you want specific extra fields or not: wdyt?

**Comment by azisyus** [comment]
@mickaelandrieu adding new fields to address form is easy part and your way is good one, hard part is i need to show them in templates/customer/addresses.tpl as separated columns i mean delivery addresses will be list in left side and  invoice addresses will be list in right side, in crud page they should show their own form and i dont think we can put this kind of feature that easily with this architecture because there is only one form and we choose delivery and invoice addresses between all  in checkoutAddressStep and everything is turning around of this
step0001

**Comment by mickaelandrieu** [comment]
I'm not really an expert of legacy form rendering, but I guess we can override the template for a specific form.

I guess @Quetzacoalt91 or @rGaillard are far better than me at answering this :+1:

**Review by matks — CHANGES_REQUESTED** [review]
I agree with @mickaelandrieu. Most of the time, modifying the scope of a variable from private to protected is a bad solution to a good problem. We should not merge this PR like this, but rather consider a better solution.

**Comment by PierreRambaud** [comment]
Hi @azisyus ,

This kind of change is already being discussed in #11910. As this one is a duplicate of the second link, I'm closing it as a duplicate. I let you participate in the debate if needed.

Regards

**Pull request closed without merging**